#159204723 Users should be able to comment on articles

madeofhuman opened the pull request:
#### What does this PR do?
- add the feature to enable users to comment on articles, reply to comments, update their comments, delete their comments, and view all comments and replies on article

#### Description of Task to be completed?
- create comment model and migrations
- add tests for the functionalities added
- create comment controller for creating a comment, getting all comments, getting a comment, updating a comment, and deleting a comment
- one-level-nest comment replies

#### How should this be manually tested?
 - clone the repo: Run `git clone https://github.com/andela/elven-ah.git` in the folder where you want the repo to be saved.
 - navigate into the newly created folder and install the dependencies: `cd elven-ah && npm install`
 - copy .env.example to .env: `cp .env.example .env`
 - provision a PostgreSQL database and add the credentials to the .env file.
 - start the server using `npm start`
 - create an account and create an article or seed some articles
 - to create a comment, make a POST request to `http://localhost:3000/api/articles/:slug/comments` using postman (Note: slug refers to the article's slug)
 - to get all comments of an article, make a GET request to `http://localhost:3000/api/articles/:slug/comments` using postman (Note: slug refers to the article's slug)
 - to get a single comment of an article, make a GET request to `http://localhost:3000/api/articles/:slug/comments/:id` using postman (Note: slug refers to the article's slug, and id refers to the id of the comment you want to retrieve)
 - to updated a comment, make a PUT request to `http://localhost:3000/api/articles/:slug/comments/:id` using postman (Note: slug refers to the article's slug, and id refers to the id of the comment you want to update)
 - to delete a comment, make a DELETE request to `http://localhost:3000/api/articles/:slug/comments/:id` using postman (Note: slug refers to the article's slug, and id refers to the id of the comment you want to delete)

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#159204723](https://www.pivotaltracker.com/story/show/159204723)

#### Screenshots (if appropriate)
<img width="1319" alt="screen shot 2018-08-13 at 12 43 19 pm" src="https://user-images.githubusercontent.com/7345686/44029842-b74f18c0-9ef6-11e8-9284-947e205c5416.png">

#### Questions:
N/A

[Finishes #159204723]

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
add social login id to user table

Commit:
feat(article): Create, Update and Delete Article
- Authenticated user should be able to create an article
- Authenticated user should be able to update an article
- Authenticated user should be able to delete an article
- Unit test for each features
[Delivers #159204721]

Commit:
make slug lower case and stringify date string

Commit:
modify comment and reply models and migrateion

madeofhuman committed:
feat(comment): add comments to articles

- fix user, article, and comment models and associations
- add createComment
- add getComment
- add comment validation helper

[Delivers #159204723]

madeofhuman committed:
feat(comment): add comments to articles

- fix user, article, and comment models and associations
- add createComment
- add getComment
- add comment validation helper
- add tests

[Delivers #159204723]

madeofhuman committed:
chore(tests): add more tests

  - add more tests

[Delivers #159204723]

madeofhuman committed:
chore(tests): rebase on develop

  - rebase on develop to get the changes

[Delivers #159204723]

madeofhuman committed:
bug(routes): fix comment route after rebase
  - fix breaking comment route after rebase due to change in
article route implementation

[Fixes #159204723]

madeofhuman committed:
bug(seed): remove duplicate user seed file
- remove duplicate demo-user seed file from rebase
[Fixes #159204723]

madeofhuman committed:
chore(refactor): refactor comment model and controller
  - refactor comment model and controller to use
articleId and userId as foreignKeys
[Delivers #159204723]

Veraclins reviewed (changes requested):
Good job here @madeofhuman. You did a really nice job.  just have a few areas you may want to look into.
1. Is there a reason you are having a describe block for each test case? I believe all the tests related to a particular operation in the same describe block.
2. Is there a particular reason you are using Object.assign on line 55? I believe comment returned on line 54 is what you are assigning to the constant on line 55, so why are you remapping? To be honest, I can only assume what you are doing here especially since this is the only place you are doing it.

madeofhuman commented:
@Veraclins thanks for taking your time to review my work. I really appreciate that. To address your concerns:

1. I have an overall `describe` block for my comment tests, inside of which I created _properly defined_ describe blocks for each test case I'm running. I did it like this to make the tests are well-defined, and clean.

2. I'm receiving an array of comments from the db, which contains data I don't need, so I'm extracting the data I need from them and performing some operations on it (like formatting the date) before returning to the user.

Veraclins reviewed (commented):
@madeofhuman, I am not saying that using Object.assign is wrong or that there is a problem with the code, I just wanted to understand why you are using it. Thanks for explaining that. I now get the reason you need it. Good job.

Oluwafayokemi reviewed (approved):
Good Job! CJ, Weldone

yomigeek commented on server/controllers/CommentController.js:
```diff
@@ -0,0 +1,228 @@
+          .then(([, [updatedComment]]) => {
+            res.status(200).json({
+              status: 'success',
+              data: {
```
@madeofhuman Nice work! I must implore your creativity around the approach you took regarding this task, however, based on the reference on the readme, the response returned should be "comment" and not "data".

madeofhuman replied:
Nice catch! I'll update that. Thank you.

chukwuemekachm reviewed (commented):
I'm really impressed with this @madeofhuman, well done. However please try to validate the id parameters in the URLs. This would ensure that negative values, decimals and other characters that might cause a database crash don't get to the database.
Nice job.

madeofhuman commented:
I'm glad you noticed, that, @chukwuemekachm. I will take care of that in the `catch` blocks.

madeofhuman committed:
fix(refactor): refactor code based on feedback
  - change `data` property in `getComment` response to `comment`
  - test more edge cases
[Fixes #159204723]

madeofhuman committed:
fix(syntax): fix Hound syntax error
  - remove unused auto-generated import
[Fixes #159204723]
